web.Dockerfile

```diff
@@ -21,11 +21,13 @@ RUN swift build \
 # Run image
 # ================================
 FROM vapor/ubuntu:18.04
-WORKDIR /run
+WORKDIR /app
+
+ENV PORT 80
 
 # Copy build artifacts
-COPY --from=build /build/.build/release /run
+COPY --from=build /build/.build/release /app
 # Copy Swift runtime libraries
 COPY --from=build /usr/lib/swift/ /usr/lib/swift/
 
-ENTRYPOINT ["./Run", "serve", "--env", "production", "--hostname", "0.0.0.0", "--port", "80"]
+CMD ./Run serve --env production --hostname 0.0.0.0 --port $PORT
```

Comment:
This one's a bit of a mixed bag. An image set up this way can be spawned into with a bash shell and such, yes - but the process with PID 1 will be a `/bin/sh` instance, _not_ the executed Vapor app. This means signals like `SIGINT` and `SIGSTOP` won't show up - and potentially more importantly, `SIGTERM` will _not_ be delivered when the container is `stop`'d. To get the best of both worlds, try this:
```dockerfile
ARG env
ENV env ${env:-production} # might as well

ENTRYPOINT ["./Run"]
CMD ["serve", "--env", "$env", "--hostname", "0.0.0.0", "--port", "$PORT"]
```
This way:
1. `Run` gets to be PID 1 by default.
2. The parameters passed to `Run` can be overridden by just appending them to the `docker run` command line. This is already used by `template`'s `Dockerfile`.
3. To get a bash entry point, pass `-ti --entrypoint /usr/bin/bash` to your `docker run` command; it will override both the default value and the entire CMD.

Reply:
+1 to this. From a usability standpoint I think having the Vapor executable as the entrypoint makes sense because you can directly pass commands like `migrate` and `serve` to `docker run`. `--entrypoint` can be used when you need to get into bash for some reason. Does `heroku run` allow you to pass that flag through?

Reply:
I don't think it does:

```
$ heroku run --help
run a one-off process inside a heroku dyno

USAGE
  $ heroku run

OPTIONS
  -a, --app=app        (required) app to run command against
  -e, --env=env        environment variables to set (use ';' to split multiple vars)
  -r, --remote=remote  git remote of app to use
  -s, --size=size      dyno size
  -x, --exit-code      passthrough the exit code of the remote command
  --no-notify          disables notification when dyno is up (alternatively use HEROKU_NOTIFICATIONS=0)
  --no-tty             force the command to not run in a tty
  --type=type          process type

DESCRIPTION
  Shows a notification if the dyno takes more than 20 seconds to start.

EXAMPLES
  $ heroku run bash
  Running bash on app.... up, run.1
  ~ $

  $ heroku run -s hobby -- myscript.sh -a arg1 -s arg2
  Running myscript.sh -a arg1 -s arg2 on app.... up, run.1

COMMANDS
  run:detached  run a detached dyno, where output is sent to your logs
```

...but this command predates the container runtime.

[This](https://devcenter.heroku.com/articles/container-registry-and-runtime#dockerfile-commands-and-runtime) is what I could find about the Dockerfile limitations on a whim. Doesn't help much either.